### PR TITLE
A few improvements to the PETSc classes

### PR DESCRIPTION
**Pull request by stefanozampini**
This set of self-contained commits provides a few improvements to the various PETSc classes in deal.II that I have made to design interfaces for PETSc SNES and TS solvers, respectively, nonlinear equations and time-dependent differential algebraic equation solvers.

All the new functionality provided in this MR is tested. 
PETSc SNES and TS tutorials are in step-86 and step-67 respectively

### CODE REVIEW

**Comment by peterrum** [comment]
> to design interfaces for PETSc SNES and TS solvers, respectively, nonlinear equations and time-dependent differential algebraic equation solvers. 

> open a second MR with the new solvers.

Very nice. Looking forward to the new solvers!

**Comment by stefanozampini** [comment]
> > to design interfaces for PETSc SNES and TS solvers, respectively, nonlinear equations and time-dependent differential algebraic equation solvers.
> 
> > open a second MR with the new solvers.
> 
> Very nice. Looking forward to the new solvers!

I have just finished polishing tests and tutorials. If you want to give early feedback, take a look here https://github.com/luca-heltai/dealii/commits/petsc-ts and comment directly there

**Comment by peterrum** [comment]
@stefanozampini I left two comments on your branch. Generally, what about creating a big PR with everything? You can still split up the PR in smaller PRs but it would make it easier to get an overview and to discuss.

**Comment by luca-heltai** [comment]
I was the one suggesting smaller PRs... :D Go ahead, and throw everything in, @stefanozampini ! :)

**Comment by luca-heltai** [comment]
(by the way, 5b89b32 is a dumb bug in step-77... I'm not sure how it could work before...)

**Comment by stefanozampini** [comment]
Ok, I have moved all the commits here to ease discussion. At least to me, these are the topics I would like to get your feedback on:

- To limit unneeded memory copies, we probably need to augment the PETSc Vector classes inheriting from `VectorSpaceVector`. The SNES tutorial is ok, but the Runge-Kutta tutorial copies input-output in its current version.
- How to provide source code customization of specific solver subtypes via parameters? Currently, I only get the solver type; e.g `rk` for Runge-Kutta, but the specific subtype cannot be selected via parameters.
- PETSc support KOKKOS, interface code from deal.II is missing but could be easily provided.
- I'm not a cmake fan, and someone needs to write some extra code to better integrate `PETSC_HAVE_KOKKOS` into deal.II configure system
- CUDA/HIP support is quite mature in PETSc, this can also be supported from deal.II

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 1]
```diff
@@ -295,12 +295,19 @@ namespace PETScWrappers
...
+                                              colnums_in_window.data(),
+                                              nullptr);
           AssertThrow(ierr == 0, ExcPETScError(ierr));
         }
       else
         {
           PetscInt i = 0;
-          ierr       = MatMPIAIJSetPreallocationCSR(matrix, &i, &i, nullptr);
+
+          ierr = MatSeqAIJSetPreallocationCSR(matrix, &i, &i, nullptr);
```
Comment (knepley) [inline source/lac/petsc_parallel_sparse_matrix.cc:308]: Could use MatXAIJSetPreallocation() here

**Comment by bangerth** [comment]
Having all of these commits as part of one pull request makes it more cumbersome than necessary to get things merge. Please break things into logically self-contained subsets. For example, the doc changes and the step-77 thing (duh!) can easily be merged and be moved out of the way already.

For tutorial programs, please also see #11998. step-73 is already reserved. You might want to add your proposed steps there as well.

**Comment by stefanozampini** [comment]
> Having all of these commits as part of one pull request makes it more cumbersome than necessary to get things merge. Please break things into logically self-contained subsets. For example, the doc changes and the step-77 thing (duh!) can easily be merged and be moved out of the way already.

@bangerth Each commit is self-contained, and it compiles. Every commit addresses one issue at a time. There are no commits fixing intermediate state errors. Review could be done by commits. I maintained the old behavior of all the classes I have touched.

**Comment by stefanozampini** [comment]
> For tutorial programs, please also see #11998. step-73 is already reserved. You might want to add your proposed steps there as well.

@bangerth Ok then, I will reserve step-86 then. I will rename my step-73 once this gets approved if it is ok with you.

**Comment by stefanozampini** [comment]
@bangerth The main point of these commits is always to have a `Vec` or a `Mat` stored inside the deal.II PETSc classes. This allows us to use `SNES` and `TS` without copying data around. A complete no-copy implementation of step 67 will require more work to implement the `VectorSpaceVector` methods, but this could be left for a second PR.

**Comment by luca-heltai** [comment]
Opened separate PR for some of the commits:

#14575 

#14592

#14591

**Comment by luca-heltai** [comment]
#14601

**Comment by peterrum** [comment]
@stefanozampini Sorry for getting back so late. Nice additions!

@vovannikov and I have been working on the integration of the new SNES wrapper into our application code (a matrix-free code with our own block vectors that are build around deal.II's `LinearAlgebra::distributed::Vector`) as alternative to (the now merged) NOX wrapper (https://github.com/dealii/dealii/blob/master/include/deal.II/trilinos/nox.h).

 For that purpose, we wrote a wrapper around your wrapper, which resembles more the one of the NOX wrapper:

```cpp
#pragma once

#include <pf-applications/lac/dynamic_block_vector.h>

#ifdef DEAL_II_WITH_PETSC

#  include <deal.II/lac/petsc_block_sparse_matrix.h>
#  include <deal.II/lac/petsc_block_vector.h>
#  include <deal.II/lac/petsc_snes.h>
#  include <deal.II/lac/petsc_vector.h>

namespace NonLinearSolvers
{
  using namespace dealii;

  DeclException0(ExcSNESNoConvergence);

  template <typename VectorType>
  struct PETSCVectorTraits;

  template <typename Number>
  struct PETSCVectorTraits<LinearAlgebra::distributed::Vector<Number>>
  {
    using VectorType  = LinearAlgebra::distributed::Vector<Number>;
    using PVectorType = PETScWrappers::MPI::Vector;
    using PMatrixType = PETScWrappers::MatrixBase; // not needed!?

    static void
    copy(VectorType &dst, const PVectorType &src)
    {
      for (const auto i : dst.locally_owned_elements())
        dst[i] = src[i];
    }

    static void
    copy(PVectorType &dst, const VectorType &src)
    {
      dst = 0.0;
      for (const auto i : src.locally_owned_elements())
        dst[i] = src[i];
      dst.compress(VectorOperation::insert);
    }

    static std::shared_ptr<PVectorType>
    create(const VectorType &vector)
    {
      return std::make_shared<PVectorType>(vector.locally_owned_elements(),
                                           vector.get_mpi_communicator());
    }
  };

  template <typename Number>
  struct PETSCVectorTraits<
    LinearAlgebra::distributed::DynamicBlockVector<Number>>
  {
    using VectorType  = LinearAlgebra::distributed::DynamicBlockVector<Number>;
    using PVectorType = PETScWrappers::MPI::BlockVector;
    using PMatrixType = PETScWrappers::MPI::BlockSparseMatrix; // not needed!?

    static void
    copy(VectorType &dst, const PVectorType &src)
    {
      dst = 0.0;
      for (unsigned int b = 0; b < dst.n_blocks(); ++b)
        for (const auto i : dst.block(b).locally_owned_elements())
          dst.block(b)[i] = src.block(b)[i];
    }

    static void
    copy(PVectorType &dst, const VectorType &src)
    {
      dst = 0.0;
      for (unsigned int b = 0; b < src.n_blocks(); ++b)
        for (const auto i : src.block(b).locally_owned_elements())
          dst.block(b)[i] = src.block(b)[i];
      dst.compress(VectorOperation::insert);
    }

    static std::shared_ptr<PVectorType>
    create(const VectorType &vector)
    {
      std::vector<IndexSet> index_sets(vector.n_blocks());

      for (unsigned int b = 0; b < vector.n_blocks(); ++b)
        index_sets[b] = vector.block(b).locally_owned_elements();

      auto new_vector =
        std::make_shared<PVectorType>(index_sets,
                                      vector.block(0).get_mpi_communicator());

      new_vector->collect_sizes();

      return new_vector;
    }
  };

  template <typename VectorType>
  class SNESSolver
  {
  private:
    using VectorTraits = PETSCVectorTraits<VectorType>;
    using PVectorType  = typename VectorTraits::PVectorType;
    using PMatrixType  = typename VectorTraits::PMatrixType; // not needed!?

  public:
    struct AdditionalData
    {
    public:
      AdditionalData(const unsigned int max_iter                       = 10,
                     const double       abs_tol                        = 1.e-20,
                     const double       rel_tol                        = 1.e-5,
                     const unsigned int threshold_nonlinear_iterations = 1,
                     const unsigned int threshold_n_linear_iterations  = 0,
                     const std::string &line_search_type               = "bt",
                     const bool         reuse_solver                   = false);

      unsigned int max_iter;
      double       abs_tol;
      double       rel_tol;
      unsigned int threshold_nonlinear_iterations;
      unsigned int threshold_n_linear_iterations;
      std::string  line_search_type;
      bool         reuse_solver;
    };

    SNESSolver(const AdditionalData &additional_data,
               const std::string &   snes_type = "");

    void
    clear();

    unsigned int
    solve(VectorType &vector);

    std::function<int(const VectorType &x, VectorType &f)> residual;

    std::function<int(const VectorType &x)> setup_jacobian;

    std::function<int(const VectorType &x)> setup_preconditioner;

    // TODO: not uses yet
    std::function<int(const VectorType &x, VectorType &v)> apply_jacobian;

    std::function<
      int(const VectorType &f, VectorType &x, const double tolerance)>
      solve_with_jacobian;

    std::function<
      int(const VectorType &f, VectorType &x, const double tolerance)>
      solve_with_jacobian_and_track_n_linear_iterations;

    std::function<SolverControl::State(const unsigned int i,
                                       const double       norm_f,
                                       const VectorType & x,
                                       const VectorType & f)>
      check_iteration_status;

    std::function<bool()> update_preconditioner_predicate;

  private:
    AdditionalData additional_data;

    const std::string snes_type;

    unsigned int n_residual_evaluations;
    unsigned int n_jacobian_applications;
    unsigned int n_nonlinear_iterations;
    unsigned int n_last_linear_iterations;
  };



  template <typename VectorType>
  SNESSolver<VectorType>::AdditionalData::AdditionalData(
    const unsigned int max_iter,
    const double       abs_tol,
    const double       rel_tol,
    const unsigned int threshold_nonlinear_iterations,
    const unsigned int threshold_n_linear_iterations,
    const std::string &line_search_type,
    const bool         reuse_solver)
    : max_iter(max_iter)
    , abs_tol(abs_tol)
    , rel_tol(rel_tol)
    , threshold_nonlinear_iterations(threshold_nonlinear_iterations)
    , threshold_n_linear_iterations(threshold_n_linear_iterations)
    , line_search_type(line_search_type)
    , reuse_solver(reuse_solver)
  {}



  template <typename VectorType>
  SNESSolver<VectorType>::SNESSolver(const AdditionalData &additional_data,
                                     const std::string &   snes_type)
    : additional_data(additional_data)
    , snes_type(snes_type)
    , n_residual_evaluations(0)
    , n_jacobian_applications(0)
    , n_nonlinear_iterations(0)
    , n_last_linear_iterations(0)
  {}



  template <typename VectorType>
  void
  SNESSolver<VectorType>::clear()
  {
    // clear interal counters
    n_residual_evaluations   = 0;
    n_jacobian_applications  = 0;
    n_nonlinear_iterations   = 0;
    n_last_linear_iterations = 0;
  }



  template <typename VectorType>
  unsigned int
  SNESSolver<VectorType>::solve(VectorType &vector)
  {
    if (additional_data.reuse_solver == false)
      clear(); // clear state

    typename PETScWrappers::NonlinearSolverData p_additional_data;

    p_additional_data.snestype           = snes_type;
    p_additional_data.absolute_tolerance = additional_data.abs_tol;
    p_additional_data.relative_tolerance = additional_data.rel_tol;
    p_additional_data.max_it             = additional_data.max_iter;

    PETScWrappers::NonlinearSolver<PVectorType, PMatrixType> solver(
      p_additional_data);

    if (additional_data.line_search_type.size())
      {
        SNESLineSearch linesearch;
        SNESGetLineSearch(solver.petsc_snes(), &linesearch);
        AssertSNES(
          SNESLineSearchSetType(linesearch,
                                additional_data.line_search_type.c_str()));
      }

    // create a temporal PETSc vector
    auto pvector = VectorTraits::create(vector);

    // create temporal deal.II vectors
    VectorType tmp_0, tmp_1;
    tmp_0.reinit(vector);
    tmp_1.reinit(vector);

    // wrap deal.II functions
    solver.residual = [&](const PVectorType &X, PVectorType &F) -> int {
      Assert(
        residual,
        ExcMessage(
          "No residual function has been attached to the SNESSolver object."));

      n_residual_evaluations++;

      VectorTraits::copy(tmp_0, X);
      VectorTraits::copy(tmp_1, F);
      const auto flag = this->residual(tmp_0, tmp_1);
      VectorTraits::copy(F, tmp_1);
      return flag;
    };

    solver.setup_jacobian = [&](const PVectorType &X) -> int {
      Assert(
        setup_jacobian,
        ExcMessage(
          "No setup_jacobian function has been attached to the SNESSolver object."));

      VectorTraits::copy(tmp_0, X);

      auto flag = this->setup_jacobian(tmp_0);

      if (flag != 0)
        return flag;

      if (setup_preconditioner)
        {
          // check if preconditioner needs to be updated
          bool update_preconditioner =
            ((additional_data.threshold_nonlinear_iterations > 0) &&
             ((n_nonlinear_iterations %
               additional_data.threshold_nonlinear_iterations) == 0)) ||
            (solve_with_jacobian_and_track_n_linear_iterations &&
             (n_last_linear_iterations >
              additional_data.threshold_n_linear_iterations));

          if ((update_preconditioner == false) &&
              (update_preconditioner_predicate != nullptr))
            update_preconditioner = update_preconditioner_predicate();

          if (update_preconditioner) // update preconditioner
            flag = setup_preconditioner(tmp_0);
        }

      return flag;
    };

    solver.solve_for_jacobian_system = [&](const PVectorType &src,
                                           PVectorType &      dst) -> int {
      // TODO: tolerance not used
      const double tolerance = 0.0;

      n_nonlinear_iterations++;

      // invert Jacobian
      if (solve_with_jacobian)
        {
          Assert(
            !solve_with_jacobian_and_track_n_linear_iterations,
            ExcMessage(
              "It does not make sense to provide both solve_with_jacobian and "
              "solve_with_jacobian_and_track_n_linear_iterations!"));

          // without tracking of linear iterations
          VectorTraits::copy(tmp_0, src);
          VectorTraits::copy(tmp_1, dst);
          const auto flag = solve_with_jacobian(tmp_0, tmp_1, tolerance);
          VectorTraits::copy(dst, tmp_1);

          return flag;
        }
      else if (solve_with_jacobian_and_track_n_linear_iterations)
        {
          // with tracking of linear iterations
          VectorTraits::copy(tmp_0, src);
          VectorTraits::copy(tmp_1, dst);
          const int n_linear_iterations =
            solve_with_jacobian_and_track_n_linear_iterations(tmp_0,
                                                              tmp_1,
                                                              tolerance);
          VectorTraits::copy(dst, tmp_1);

          if (n_linear_iterations == -1)
            return 1;

          this->n_last_linear_iterations = n_linear_iterations;

          return 0;
        }
      else
        {
          Assert(false,
                 ExcMessage(
                   "Neither a solve_with_jacobian or a "
                   "solve_with_jacobian_and_track_n_linear_iterations function "
                   "has been attached to the SNESSolver object."));

          Assert(false, ExcNotImplemented());
          return 1;
        }
    };

    if (check_iteration_status)
      solver.monitor = [&](const PVectorType &x,
                           const unsigned int step_number,
                           double             fval) -> int {
        VectorTraits::copy(tmp_0, x);

        if (false)
          {
            this->residual(tmp_0, tmp_1);
            this->check_iteration_status(step_number, fval, tmp_0, tmp_1);
          }
        else
          {
            this->check_iteration_status(step_number,
                                         fval,
                                         tmp_0,
                                         VectorType());
          }

        return 0;
      };

    // solve
    VectorTraits::copy(*pvector, vector);
    const unsigned int n_iterations = solver.solve(*pvector);
    VectorTraits::copy(vector, *pvector);

    AssertThrow(additional_data.max_iter == 0 ||
                  n_iterations < additional_data.max_iter,
                ExcSNESNoConvergence());

    return n_iterations;
  }

} // namespace NonLinearSolvers

#endif
```

While doing so, we noticed following:
- newer version of PETSc was needed (I switched from 3.10.5 to master)
- `MatrixType` is not always needed so that it is a bit odd that one has to specify it in the matrix-free case
- copy between deal.II and PETSc vectors is needed (see first 100 locs)
- `solve_for_jacobian_system` -> `solve_with_jacobian` -> see Kinsol and NOX wrappers
- `solve_with_jacobian` normally takes a tolerance -> see Kinsol and NOX wrappers
- no assert is thrown when max iterations is reached

Overall, the SNES wrapper seems to work. We can simply switch between nonlinear-solver types and line-search type :+1:  One combination that did not work is, when we set `ngrmes` + `bt`. We get the following error:
```
[0]PETSC ERROR: Object is in wrong state
[0]PETSC ERROR: MatMFFDSetBase() has not been called, this is often caused by forgetting to call 
		MatAssemblyBegin/End on the first Mat in the SNES compute function
[0]PETSC ERROR: WARNING! There are option(s) set that were not used! Could be the program crashed before they were used or a spelling mistake, etc!
[0]PETSC ERROR: Option left: name:--circle value: 15
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.18.3, unknown 
[0]PETSC ERROR: ./applications/sintering/sintering-2D-generic-scalar on a arch-linux-c-debug named node29.cluster by munch Tue Jan  3 08:29:23 2023
[0]PETSC ERROR: Configure options 
[0]PETSC ERROR: #1 MatMult_MFFD() at /home/munch/sw-sintering/petsc/src/mat/impls/mffd/mffd.c:312
[0]PETSC ERROR: #2 MatMult_Shell() at /home/munch/sw-sintering/petsc/src/mat/impls/shell/shell.c:1014
[0]PETSC ERROR: #3 MatMult() at /home/munch/sw-sintering/petsc/src/mat/interface/matrix.c:2570
[0]PETSC ERROR: #4 SNESLineSearchApply_BT() at /home/munch/sw-sintering/petsc/src/snes/linesearch/impls/bt/linesearchbt.c:125
[0]PETSC ERROR: #5 SNESLineSearchApply() at /home/munch/sw-sintering/petsc/src/snes/linesearch/interface/linesearch.c:607
[0]PETSC ERROR: #6 SNESSolve_NGMRES() at /home/munch/sw-sintering/petsc/src/snes/impls/ngmres/snesngmres.c:243
[0]PETSC ERROR: #7 SNESSolve() at /home/munch/sw-sintering/petsc/src/snes/interface/snes.c:4692
```

I guess the backtracking algorithm wants to apply the Jacobian; however, this is not possible with the current interface. For the NOX-wrapper, we had to add `std::function<int(const VectorType &x, VectorType &v)> apply_jacobian;` for this purpose.

**Comment by stefanozampini** [comment]
@peterrum, thanks for the feedback.

> newer version of PETSc was needed (I switched from 3.10.5 to master)

3.10.5 is from 3 years ago, and we have the bad habit of changing API quite often in such a period of time. If you need to use that version, can you send me the compilation errors? Also, what is the minimal PETSc version you would like to support from deal.ii?
 `master` is frozen. To track the dev version you must now use `main`. For the release + patches instead, track `release`

> MatrixType is not always needed so that it is a bit odd that one has to specify it in the matrix-free case

That is there if you want to use the pure PETSc interface, which is not your case

> copy between deal.II and PETSc vectors is needed (see first 100 locs)

The SNES example in this MR does not use copies. Work is needed to provide better integration of the PETSc vectors with the VectorSpaceVectors or other methods you may need with your derived classes. I'm no deal.II expert, and I would welcome any help in that direction.

> solve_for_jacobian_system -> solve_with_jacobian -> see Kinsol and NOX wrappers

That was @luca-heltai suggestion. I can change the naming convention

> solve_with_jacobian normally takes a tolerance -> see Kinsol and NOX wrappers

`solve_with_jacobian` is used as a preconditioner inside the PETSc classes to allow for approximate solvers defined by the user and possibly iterate on those. It does not make sense to provide a tolerance. At some point, one has to throw a line between what a user must supply and what can be done automatically

> no assert is thrown when max iterations is reached

If there's consensus on what to do in which situation, I'll be glad to implement it

**Comment by bangerth** [comment]
I believe that everything in this PR is now merged, in some form or other. (We decided to put the two modified tutorial programs into the test suite, and they now are.)

@stefanozampini Do you agree? If so, can we close this PR?

**Comment by stefanozampini** [comment]
sure, everything has been merged

**Pull request closed without merging**